### PR TITLE
WIP: Skip over YouTube post-live manifestless mode videos during import

### LINKED ISSUE #5331: Skip over post-processing and still-to-be-premiered videos
### Describe the current behavior

PeerTube currently syncs videos that are still being post-processed by YouTube.

For livestreams with a length over 2 hours that finished broadcasting, but haven't finished post-processing this results in [incomplete and cut off videos](https://github.com/yt-dlp/yt-dlp/issues/1564) (post-live manifestless mode) to be imported.

### Steps to reproduce

1. Setup PeerTube using latest develop branch (2022-10-07)
2. Synchronize a YouTube channel
3. Wait for or publish a livestream that exceeds 2 hours in length on YouTube
4. Kick off the synchronization after the livestream finishes
5. Observe PeerTube importing a incomplete and cut-off video

### Describe the expected behavior

PeerTube only importing the video when YouTube has fully processed it, skipping over videos that are still in [post-live manifestless mode](https://github.com/yt-dlp/yt-dlp/commit/829bbd1d05ae7e4519327f1cb6e75b3da38e0a0b).

### Additional information

* PeerTube instance:
  * URL: https://pcbu.nl
  * Version: latest code in develop branch (2022-10-07)
  * Node.js version: v16.17.1
  * FFmpeg version: 4.3.4-0+deb11u1

### PR DESCRIPTION
**Pull request by emansom**
🚧👷 **Do not merge** Refactoring.

# Depends on
✔️ ~~- [yt-dlp](https://github.com/yt-dlp/yt-dlp) releasing a new version that includes fixes related to fetching `live_status`.~~

## Description

PeerTube currently syncs videos that are still being post-processed by YouTube. This results in [incomplete and cut off videos](https://github.com/yt-dlp/yt-dlp/issues/1564) (post-live manifestless mode) to be imported.

The code in this PR resolves that issue.

## Changes

- Videos in post-live manifestless mode are filtered out from `youtube-dl` playlists and direct retrieval by URL will error out.
- Seperated out the `YoutubeDLInfoBuilder` to a generic base builder and implemented a builder for `yt-dlp` playlist data.
- Added `youtube-dl` tag to various log calls to aid in debugging.
- Added `channel-sync` tag to the log calls inside the channel synchronization task.
- Implemented TypeScript types for `youtube-dl`'s JSON-LD output.
- Add type hints to functions inside YoutubeDLInfoBuilder.
- Remove redundant null check in isNSFW

## Related issues

Fixes #5331

## Has this been tested?

Unsure how we can test this reliably in a programmatic matter using unit tests, as it relies on YouTube's transcoding behavior in long livestreams.

It has however been tested in production on [pcbu.nl](https://pcbu.nl/)

- [ ] 👍 yes, I added tests to the test suite
- [ ] 💭 no, because this PR is a draft and still needs work
- [ ] 🙅 no, because this PR does not update server code
- [ ] 🙋 no, because I need help

### CODE REVIEW

**Comment by Chocobozzz** [comment]
No need for tests for this :)

**Comment by emansom** [comment]
- `--ignore-no-formats-error` is passed to `youtube-dl` to make it not crash during premieres/livestreams, check logic on `.formats` has been implemented
- Videos in post-live manifestless mode are filtered out from `yt-dlp` playlists and direct retrieval by URL will error out.
- Seperated out the `YoutubeDLInfoBuilder` to a generic base builder and implemented a builder for `yt-dlp` playlist data.
- Added `youtube-dl` tag to various log calls to aid in debugging.

**Comment by emansom** [comment]
@Chocobozzz Removed WIP tag, safe to merge now.

**Comment by emansom** [comment]
@Chocobozzz Nevermind 😅 

There's a `async` prefix introduced to the `getListInfo` function in commit 5cc2f0ea0a0af82333d673db1ad6461ed5cb4233, will alter the code to reflect now.

**Comment by emansom** [comment]
> @Chocobozzz Nevermind sweat_smile
> 
> There's a `async` prefix introduced to the `getListInfo` function in commit [5cc2f0e](https://github.com/Chocobozzz/PeerTube/commit/5cc2f0ea0a0af82333d673db1ad6461ed5cb4233), will test again.

@Chocobozzz No conflicts with https://github.com/Chocobozzz/PeerTube/commit/5cc2f0ea0a0af82333d673db1ad6461ed5cb4233. Safe to merge. 👍🏻

File: server/helpers/youtube-dl/youtube-dl-cli.ts [hunk 1]
```diff
@@ -139,7 +138,7 @@ export class YoutubeDLCLI {
     url: string
     latestVideosCount?: number
     processOptions: execa.NodeOptions
-  }): Promise<{ upload_date: string, webpage_url: string }[]> {
```
Comment (Chocobozzz) [inline server/helpers/youtube-dl/youtube-dl-cli.ts:142]: Why did you remove the typing information?

File: server/helpers/youtube-dl/youtube-dl-base-builder.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+  VIDEO_CATEGORIES,
+  VIDEO_IMPORT_LIVE_STATUSES,
+  VIDEO_LANGUAGES,
+  VIDEO_LICENCES
+} from '@server/initializers/constants'
+import { peertubeTruncate } from '@server/helpers/core-utils'
+import { isUrlValid } from '@server/helpers/custom-validators/activitypub/misc'
+
+class YoutubeDLBaseBuilder {
+  public normalizeObject (obj: any) {
```
Comment (Chocobozzz) [inline server/helpers/youtube-dl/youtube-dl-base-builder.ts:12]: Please remove the `public` keyword for consistency

File: server/helpers/youtube-dl/youtube-dl-cli.ts [hunk 3]
```diff
@@ -105,7 +104,7 @@ export class YoutubeDLCLI {
     additionalYoutubeDLArgs?: string[]
   }) {
     let args = options.additionalYoutubeDLArgs || []
-    args = args.concat([ '--merge-output-format', 'mp4', '-f', options.format, '-o', options.output ])
+    args = args.concat([ '--ignore-no-formats-error', '--merge-output-format', 'mp4', '-f', options.format, '-o', options.output ])
```
Comment (Chocobozzz) [inline server/helpers/youtube-dl/youtube-dl-cli.ts:107]: Can you add a comment to explain the purpose of `--ignore-no-formats-error'`?

File: server/helpers/youtube-dl/youtube-dl-wrapper.ts [hunk 4]
```diff
@@ -20,6 +22,32 @@ const processOptions = {
   maxBuffer: 1024 * 1024 * 30 // 30MB
 }
 
+class YoutubeDLError extends Error {
+  code: YoutubeDLError.CODE
+  cause?: Error // Property to remove once ES2022 is used
```
Comment (Chocobozzz) [inline server/helpers/youtube-dl/youtube-dl-wrapper.ts:27]: Prefix your comment with `// TODO`

File: server/helpers/youtube-dl/youtube-dl-wrapper.ts [hunk 5]
```diff
@@ -71,13 +116,17 @@ class YoutubeDLWrapper {
     const youtubeDL = await YoutubeDLCLI.safeGet()
 
     const files = await youtubeDL.getSubs({ url: this.url, format: 'vtt', processOptions: { cwd } })
-    if (!files) return []
+    if (!files) {
```
Comment (Chocobozzz) [inline server/helpers/youtube-dl/youtube-dl-wrapper.ts:119]: Please don't update code you did you modify :)

File: server/helpers/youtube-dl/youtube-dl-wrapper.ts [hunk 6]
```diff
@@ -71,13 +116,17 @@ class YoutubeDLWrapper {
...
+      return []
+    }
 
     logger.debug('Get subtitles from youtube dl.', { url: this.url, files, ...lTags() })
 
     const subtitles = files.reduce((acc, filename) => {
       const matched = filename.match(/\.([a-z]{2})(-[a-z]+)?\.(vtt|ttml)/i)
-      if (!matched || !matched[1]) return acc
+      if (!matched || !matched[1]) {
```
Comment (Chocobozzz) [inline server/helpers/youtube-dl/youtube-dl-wrapper.ts:127]: Same here

File: server/helpers/youtube-dl/youtube-dl-wrapper.ts [hunk 7]
```diff
@@ -118,7 +167,7 @@ class YoutubeDLWrapper {
     } catch (err) {
       this.guessVideoPathWithExtension(pathWithoutExtension, fileExt)
         .then(path => {
-          logger.debug('Error in youtube-dl import, deleting file %s.', path, { err, ...lTags() })
+          logger.debug(`Error in youtube-dl import, deleting file ${path}.`, { err, ...lTags() })
```
Comment (Chocobozzz) [inline server/helpers/youtube-dl/youtube-dl-wrapper.ts:170]: Why did you update this line?

File: server/helpers/youtube-dl/youtube-dl-wrapper.ts [hunk 8]
```diff
@@ -145,6 +194,57 @@ class YoutubeDLWrapper {
 
     throw new Error(`Cannot guess path of ${tmpPath}. Directory content: ${directoryContent.join(', ')}`)
   }
+
+  private filterUnavailable (list: YoutubeDLPartialEntry[]) {
+    return list.filter((info) => {
+      switch (info.liveStatus) {
+        case VideoImportLiveStatus.LIVE:
+          logger.warning('%s is currently livestreaming, filtering from import list', info.webpageUrl, lTags())
```
Comment (Chocobozzz) [inline server/helpers/youtube-dl/youtube-dl-wrapper.ts:202]: prefer `.debug` log for every log here

File: server/helpers/youtube-dl/youtube-dl-wrapper.ts [hunk 9]
```diff
@@ -145,6 +194,57 @@ class YoutubeDLWrapper {
...
+          return false
+      }
+
+      return true
+    })
+  }
+
+  private checkFormats (info: YoutubeDLInfo) {
+    if (info.formats.length === 0) {
+      logger.warning('%s has no downloadable video formats available, skipping', info.webpageUrl, lTags())
```
Comment (Chocobozzz) [inline server/helpers/youtube-dl/youtube-dl-wrapper.ts:218]: debug log

File: server/helpers/youtube-dl/youtube-dl-wrapper.ts [hunk 10]
```diff
@@ -145,6 +194,57 @@ class YoutubeDLWrapper {
...
+      throw new YoutubeDLError({
+        message: `${info.webpageUrl} has no downloadable video formats available, skipping`,
+        code: YoutubeDLError.CODE.NO_FORMATS_AVAILABLE
+      })
+    }
+  }
+
+  private checkAvailability (info: YoutubeDLInfo | YoutubeDLPartialEntry) {
```
Comment (Chocobozzz) [inline server/helpers/youtube-dl/youtube-dl-wrapper.ts:226]: Can't you share some code with `filterUnavailable`?

File: server/helpers/youtube-dl/youtube-dl-base-builder.ts [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+    for (const key of Object.keys(VIDEO_CATEGORIES)) {
+      const category = VIDEO_CATEGORIES[key]
+      if (categoryString.toLowerCase() === category.toLowerCase()) return parseInt(key, 10)
+    }
+
+    return undefined
+  }
+
+  public getLiveStatus (status: string) {
+    return VIDEO_IMPORT_LIVE_STATUSES[status] ? status : VIDEO_IMPORT_LIVE_STATUSES.not_live
```
Comment (Chocobozzz) [inline server/helpers/youtube-dl/youtube-dl-base-builder.ts:152]: I'm not sure to understand the purpose of translating the internal youtube status to plain english?

File: server/lib/sync-channel.ts [hunk 12]
```diff
@@ -36,7 +38,7 @@ export async function synchronizeChannel (options: {
 
     logger.info(
       'Fetched %d candidate URLs for sync channel %s.',
-      targetUrls.length, channel.Actor.preferredUsername, { targetUrls }
+      targetUrls.length, channel.Actor.preferredUsername, { targetUrls }, lTags()
```
Comment (Chocobozzz) [inline server/lib/sync-channel.ts:41]: `{ targetUrls, ...lTags() }`

File: server/lib/sync-channel.ts [hunk 13]
```diff
@@ -9,6 +9,8 @@ import { VideoChannelSyncState, VideoPrivacy } from '@shared/models'
 import { CreateJobArgument, JobQueue } from './job-queue'
 import { ServerConfigManager } from './server-config-manager'
 
+const lTags = loggerTagsFactory('youtube-dl')
```
Comment (Chocobozzz) [inline server/lib/sync-channel.ts:12]: Prefer to use `channel-sync` tag, it's not really a `youtube-dl` package even if it uses this module

File: shared/models/videos/import/video-import-live-status.enum.ts [hunk 14]
```diff
@@ -0,0 +1,6 @@
+export const enum VideoImportLiveStatus {
```
Comment (Chocobozzz) [inline shared/models/videos/import/video-import-live-status.enum.ts:1]: If this enum is not shared with the client, put it directly in the youtube-dl module

**Review by Chocobozzz — CHANGES_REQUESTED** [review]

File: server/helpers/youtube-dl/youtube-dl-wrapper.ts (hunk 4 above)
Comment (emansom) [reply server/helpers/youtube-dl/youtube-dl-wrapper.ts:27]: Was copied from https://github.com/Chocobozzz/PeerTube/blob/9a3a23a834679e072c15960903e05e667ef14481/server/lib/video-import.ts#L36

Will add the TODO prefix in both places.

File: server/helpers/youtube-dl/youtube-dl-cli.ts (hunk 3 above)
Comment (emansom) [reply server/helpers/youtube-dl/youtube-dl-cli.ts:107]: In the code or PR?

The latter I did https://github.com/Chocobozzz/PeerTube/pull/5332#issuecomment-1273356279

File: server/helpers/youtube-dl/youtube-dl-wrapper.ts (hunk 5 above)
Comment (emansom) [reply server/helpers/youtube-dl/youtube-dl-wrapper.ts:119]: Okay, will revert.

File: server/helpers/youtube-dl/youtube-dl-wrapper.ts (hunk 7 above)
Comment (emansom) [reply server/helpers/youtube-dl/youtube-dl-wrapper.ts:170]: To better fit with the other' log calls syntax.

File: server/helpers/youtube-dl/youtube-dl-wrapper.ts (hunk 10 above)
Comment (emansom) [reply server/helpers/youtube-dl/youtube-dl-wrapper.ts:226]: Good point. 👍🏻  Will do.

File: server/lib/sync-channel.ts (hunk 13 above)
Comment (emansom) [reply server/lib/sync-channel.ts:12]: 0fc

File: server/helpers/youtube-dl/youtube-dl-base-builder.ts (hunk 11 above)
Comment (emansom) [reply server/helpers/youtube-dl/youtube-dl-base-builder.ts:152]: Yeah me neither 😅

Was still getting to know the codebase, it's the first piece of code modified like a week ago.

Will modify to simply use direct enum values.

File: server/lib/sync-channel.ts (hunk 13 above)
Comment (emansom) [reply server/lib/sync-channel.ts:12]: Nah I get the point. Yeah, will modify.

File: server/helpers/youtube-dl/youtube-dl-cli.ts (hunk 1 above)
Comment (emansom) [reply server/helpers/youtube-dl/youtube-dl-cli.ts:142]: As it differs from what will be read, and it will be cast to a builder interface later in the pipe.

https://github.com/Chocobozzz/PeerTube/blob/437ac0c176525908114a9a98e82aeba0bb1bace5/server/helpers/youtube-dl/youtube-dl-list-builder.ts#L29

To be more similar to [`YoutubeDLCLI.getInfo()`](https://github.com/Chocobozzz/PeerTube/blob/437ac0c176525908114a9a98e82aeba0bb1bace5/server/helpers/youtube-dl/youtube-dl-cli.ts#L117) in behavior.

Should I modify both to reflect intermediate type info read by the builders?

File: server/helpers/youtube-dl/youtube-dl-cli.ts (hunk 3 above)
Comment (emansom) [reply server/helpers/youtube-dl/youtube-dl-cli.ts:107]: Yeah not sure, left over by debugging something. Will remove.

File: server/helpers/youtube-dl/youtube-dl-cli.ts (hunk 1 above)
Comment (emansom) [reply server/helpers/youtube-dl/youtube-dl-cli.ts:142]: Added type decelerations for both.

File: server/helpers/youtube-dl/youtube-dl-base-builder.ts (hunk 2 above)
Comment (emansom) [reply server/helpers/youtube-dl/youtube-dl-base-builder.ts]: Done

File: server/helpers/youtube-dl/youtube-dl-wrapper.ts (hunk 5 above)
Comment (emansom) [reply server/helpers/youtube-dl/youtube-dl-wrapper.ts]: Reverted

File: server/helpers/youtube-dl/youtube-dl-wrapper.ts (hunk 10 above)
Comment (emansom) [reply server/helpers/youtube-dl/youtube-dl-wrapper.ts:188]: DRYified it.

File: server/lib/sync-channel.ts (hunk 12 above)
Comment (emansom) [reply server/lib/sync-channel.ts]: 👍🏻

File: server/lib/sync-channel.ts (hunk 13 above)
Comment (emansom) [reply server/lib/sync-channel.ts]: Modified to `channel-sync` tag

File: shared/models/videos/import/video-import-live-status.enum.ts (hunk 14 above)
Comment (emansom) [reply shared/models/videos/import/video-import-live-status.enum.ts]: Renamed and moved.

File: server/helpers/youtube-dl/youtube-dl-base-builder.ts (hunk 11 above)
Comment (emansom) [reply server/helpers/youtube-dl/youtube-dl-base-builder.ts]: Simplified the enum.

File: server/helpers/youtube-dl/youtube-dl-wrapper.ts (hunk 6 above)
Comment (emansom) [reply server/helpers/youtube-dl/youtube-dl-wrapper.ts]: Reverted.

File: server/helpers/youtube-dl/youtube-dl-wrapper.ts (hunk 7 above)
Comment (emansom) [reply server/helpers/youtube-dl/youtube-dl-wrapper.ts]: Reverted.

File: server/helpers/youtube-dl/youtube-dl-wrapper.ts (hunk 8 above)
Comment (emansom) [reply server/helpers/youtube-dl/youtube-dl-wrapper.ts]: Done 👍🏻

File: server/helpers/youtube-dl/youtube-dl-wrapper.ts (hunk 9 above)
Comment (emansom) [reply server/helpers/youtube-dl/youtube-dl-wrapper.ts]: 👍🏻

**Comment by emansom** [comment]
@Chocobozzz Addressed all the suggested changes. Removed draft status and safe to merge once more.

**Comment by emansom** [comment]
Woops, test is failing. Will fix later today.

**Comment by emansom** [comment]
> Woops, test is failing. Will fix later today. Back to WIP.

Resolved.

@Chocobozzz Ready for review again. 👍🏻

**Comment by emansom** [comment]
Do not merge yet. Observed some issues of it still importing live broadcasts. Debugging now.

**Comment by emansom** [comment]
@Chocobozzz Ready for review again. 👍🏻

**Comment by emansom** [comment]
@Chocobozzz Ready for review again, addressed all threads with code changes 👍🏻

**Comment by emansom** [comment]
The code imported a incomplete stream again on pcbu.nl. Debugging. Do not merge.

**Comment by emansom** [comment]
[A new release of `yt-dlp` has been published](https://github.com/yt-dlp/yt-dlp/releases/tag/2022.11.11) that includes a few new features and fixes that this MR relies on. 🎉 

Still WIP, still need to rebase and refactor.

At the moment, recovering from a wild weekend, so no programming for a few days.

**Comment by emansom** [comment]
> Hi,
> 
> What is the status of this PR? :)

Long hiatus, should finish this. Will first update my instance this weekend, lacking behind half a year of updates. Then next week I'll rebase this PR.

**Comment by Chocobozzz** [comment]
Closing due to inactivity, don't hesitate to comment if you want to continue your work! :)
